pkg/artifacts/artifacts_storage.go

```diff
@@ -0,0 +1,143 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package artifacts
+
+import (
+	"fmt"
+
+	"github.com/knative/build-pipeline/pkg/apis/pipeline/v1alpha1"
+	"github.com/knative/build-pipeline/pkg/system"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+// ArtifactStorageInterface is an interface to define the steps to copy
+// an pipeline artifact to/from temporary storage
+type ArtifactStorageInterface interface {
+	GetCopyToContainerSpec(name, sourcePath, destinationPath string) []corev1.Container
+	GetCopyFromContainerSpec(name, sourcePath, destinationPath string) []corev1.Container
+	GetSecretsVolumes() []corev1.Volume
+	IsPVC() bool
+	StorageBasePath(pr *v1alpha1.PipelineRun) string
+}
+
+// InitializeArtifactStorage will check if there is there is a
+// bucket configured or create a PVC
+func InitializeArtifactStorage(pr *v1alpha1.PipelineRun, c kubernetes.Interface) (ArtifactStorageInterface, error) {
+	needPVC := false
+	configMap, err := c.CoreV1().ConfigMaps(system.Namespace).Get(v1alpha1.BucketConfigName, metav1.GetOptions{})
+	if err != nil {
+		needPVC = true
+	}
+	if configMap != nil && configMap.Data != nil {
+		if location, ok := configMap.Data[v1alpha1.BucketLocationKey]; !ok {
+			if location == "" {
+				needPVC = true
+			}
+		}
+	}
+	if needPVC {
+		err = createPVC(pr, c)
+		if err != nil {
+			return nil, err
+		}
+		return &v1alpha1.ArtifactPVC{Name: pr.Name}, nil
+	}
+
+	return NewArtifactBucketConfigFromConfigMap(configMap)
+}
+
+// GetArtifactStorage returnes the storgae interface to enable
+// consumer code to get a container step for copy to/from storage
+func GetArtifactStorage(prName string, c kubernetes.Interface) (ArtifactStorageInterface, error) {
+	configMap, err := c.CoreV1().ConfigMaps(system.Namespace).Get(v1alpha1.BucketConfigName, metav1.GetOptions{})
+	if err != nil {
+		return &v1alpha1.ArtifactPVC{Name: prName}, nil
+	}
+	return NewArtifactBucketConfigFromConfigMap(configMap)
+}
+
+// NewArtifactBucketConfigFromConfigMap creates a Bucket from the supplied ConfigMap
+func NewArtifactBucketConfigFromConfigMap(configMap *corev1.ConfigMap) (*v1alpha1.ArtifactBucket, error) {
+	c := &v1alpha1.ArtifactBucket{}
+
+	if configMap.Data == nil {
+		return c, nil
+	}
+	if location, ok := configMap.Data[v1alpha1.BucketLocationKey]; !ok {
+		c.Location = ""
+	} else {
+		c.Location = location
+	}
+	sp := v1alpha1.SecretParam{}
+	if secretName, ok := configMap.Data[v1alpha1.BucketServiceAccountSecretName]; !ok {
+		c.Secrets = nil
```

Comment:
Do the secrets need to explicitly be set to nil here? Would the default value work?

Reply:
fixed